File: swift/Sources/WalletAPI/ProximityTransport.swift
import Foundation

/// Duplex message channel to an in-person reader for ISO 18013-5 proximity presentation. The host
/// implements it over BLE / NFC / Wi-Fi Aware; the SDK drives the device-retrieval message exchange.
public protocol ProximityTransport: Sendable {
    /// Sends one framed message (e.g. SessionData) to the reader.
    func send(_ message: [UInt8]) async throws

    /// Suspends until the reader delivers the next framed message (e.g. SessionEstablishment).
    func receive() async throws -> [UInt8]

    /// Tears down the transport; idempotent.
    func close() async
}

File: kotlin/wallet-api/src/main/kotlin/com/hopae/eudi/wallet/spi/ProximityTransport.kt
package com.hopae.eudi.wallet.spi

/**
 * Duplex message channel to an in-person reader for ISO 18013-5 proximity presentation. The host
 * implements it over BLE / NFC / Wi-Fi Aware; the SDK drives the device-retrieval message exchange.
 */
interface ProximityTransport {
    /** Sends one framed message (e.g. SessionData) to the reader. */
    suspend fun send(message: ByteArray)

    /** Suspends until the reader delivers the next framed message (e.g. SessionEstablishment). */
    suspend fun receive(): ByteArray

    /** Tears down the transport; idempotent. */
    suspend fun close()

    /**
     * ISO 18013-5 `DeviceRetrievalMethod` entries (CBOR-encoded) this transport advertises, embedded into the QR
     * `DeviceEngagement` so the reader knows how to connect (e.g. the BLE service UUID). Empty = engagement carries none.
     */
    fun retrievalMethods(): List<ByteArray> = emptyList()
}
